### PR TITLE
8326227: Rounding error that may distort computeNextGaussian results

### PR DESCRIPTION
**Pull request by Pr0methean**
This provides a slightly more accurate bounding limit for `computeNextExponentialSoftCapped`. Currently, if the maximum is specified as 12.0, it won't actually return a value larger than `1.5 * exponentialX0` (11.353912041222094); and the error gets worse as we go further into the tail. (This affects slightly less than 12 outputs per million for an ideal RNG.) This could cause the `while (computeNextExponentialSoftCapped(rng, limit) < limit)` check in `computeNextGaussian` on line 1402 to always be true, making `nextGaussian` loop infinitely in the worst case (with probability $$e^{-2^{63} exponentialX0} \approx 10^{-3.022 \times 10^{19}}$$ for an ideal RNG); but more likely, it would give a result that was truncated too close to zero.

This change is being tested prior to submission to OpenJDK by https://github.com/openjdk/jdk/pull/17703/commits/b8be051cbf40a6a05fafc6a2c76942e9e0b11fdf.

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8326227](https://bugs.openjdk.org/browse/JDK-8326227): Rounding error that may distort computeNextGaussian results (**Bug** - P3)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17703/head:pull/17703` \
`$ git checkout pull/17703`

Update a local copy of the PR: \
`$ git checkout pull/17703` \
`$ git pull https://git.openjdk.org/jdk.git pull/17703/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17703`

View PR using the GUI difftool: \
`$ git pr show -t 17703`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17703.diff">https://git.openjdk.org/jdk/pull/17703.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17703#issuecomment-1954976728)

### CODE REVIEW

**Comment by jaikiran** [comment]
Hello Chris @Pr0methean, I don't have knowledge of this area to know if this is indeed a bug that is being addressed. I think it will be better to open a bug for this through https://bugs.java.com/bugdatabase/ and once it gets triaged and assigned JDK bug id then you can link this PR to that issue so that it goes through the usual review process.

**Comment by Pr0methean** [comment]
I've submitted a bug report for this issue at https://bugs.java.com/bugdatabase/view_bug.do?bug_id=9076599; it hasn't been published yet.

**Comment by jaikiran** [comment]
Hello Chris, the issue has now made it to the JDK project https://bugs.openjdk.org/browse/JDK-8326227. Please update the title of this PR to `8326227: Rounding error that may distort computeNextGaussian results` to officially trigger the review.

**Comment by jddarcy** [comment]
Is it practical to write a regression test for this situation?

**Comment by Pr0methean** [comment]
@jddarcy Added a regression test. Once some of my GitHub Actions runners are free to run it, https://github.com/Pr0methean/jdk/actions/runs/7983430797 will confirm that the test fails without the fix.

**Comment by Pr0methean** [comment]
Update: confirmed that the new test fails without the fix.

**Comment by jddarcy** [comment]
> Update: confirmed that the new test fails without the fix.

Thanks for verifying the test checks the fix; I'll let others who have worked more directly on the random code review the actual fix.

**Comment by Pr0methean** [comment]
@turbanoff @rgiulietti This is a follow-up to my previously merged #8131.

**Comment by rgiulietti** [comment]
@Pr0methean I'm not familiar with this code, nor with the underlying theory, so I'm not really able to give a meaningful review.

**Comment by Pr0methean** [comment]
Keep open

**Comment by Pr0methean** [comment]
Keep open; @JimLaskey it looks like you're the author of most of the surrounding code; can you confirm that there's a rounding error in the existing implementation and that this PR will fix it?

**Comment by jaikiran** [comment]
Hello Chris @Pr0methean, please keep this open. Some of us are checking if we can find someone experienced with this code to provide reviews.

**Comment by Pr0methean** [comment]
keep open

**Comment by jaikiran** [comment]
Hello Chris, I just noticed that there have been (unrelated) changes in this file in the master. Could you please merge the latest master branch changes into this PR?

File: src/java.base/share/classes/jdk/internal/util/random/RandomSupport.java [hunk 1]
```diff
@@ -1210,15 +1210,16 @@ public static double computeNextExponentialSoftCapped(RandomGenerator rng, doubl
...
         final long maxExtraMinus1;
-        if (maxValue >= MAX_EXPONENTIAL) {
+        if (maxValue >= Math.nextDown(MAX_EXPONENTIAL)) {
             maxExtraMinus1 = Long.MAX_VALUE;
         } else {
-            // Conversion to long rounds toward zero
-            maxExtraMinus1 = (long) (maxValue / DoubleZigguratTables.exponentialX0);
+            // Math.round rounds toward infinity in conversion to long; adding 1.0 corrects for any
+            // downward rounding error in the division
+            maxExtraMinus1 = Math.round(1.0 + maxValue / DoubleZigguratTables.exponentialX0);
```
Comment (jaikiran) [inline src/java.base/share/classes/jdk/internal/util/random/RandomSupport.java:1222]: We had asked Guy Steele for his inputs on these proposed changes. Guy reviewed these changes and for this specific if/else block change, he notes that (I've paraphrased it):

> the (proposed new) computation of `maxExtraMinus1` looks okay to me: it’s always okay for maxExtraMinus1 to be a bit larger than you might expect; the only downside is that the (for) loop starting (on the next line) might take extra iterations.

File: src/java.base/share/classes/jdk/internal/util/random/RandomSupport.java [hunk 2]
```diff
@@ -1210,15 +1210,16 @@ public static double computeNextExponentialSoftCapped(RandomGenerator rng, doubl
             return DoubleZigguratTables.exponentialX[(int)i] * (U1 >>> 1);
         }
         // We didn't use the upper part of U1 after all.  We'll probably be able to use it later.
-        if (maxValue <= 0.0) {
-            return 0.0;
+        if (maxValue <= DoubleZigguratTables.exponentialX0) {
+            return DoubleZigguratTables.exponentialX0;
```
Comment (jaikiran) [inline src/java.base/share/classes/jdk/internal/util/random/RandomSupport.java:1214]: Chris, could you explain why this change to this if block is necessary? Guy notes that (I've paraphrased):

> I can see that (without this proposed change here), if `maxValue` is not greater than `DoubleZigguratTables.exponentialX0`, then the subsequent computation for `maxExtraMinus1`, a few lines below using `Math.round()`, will compute 1 or 2 for the value of `maxExtraMinus1`. But with this proposed change, it just returns `DoubleZigguratTables.exponentialX0`, which by the contract of `computeNextExponentialSoftCapped` is okay (the doc says "{@code maxValue} is a "soft" cap in that a value larger than {@code maxValue} may be returned in order to save a calculation" and I remember that that is true). So that is also okay. The motivation for doing so would be that it saves computation time overall. The part I don't quite understand yet is the judgment that it will in fact save computation time overall. If you take advantage of the "soft cap" then it could cause additional iteration of the loop in `computeNextGaussian` where this `computeNextExponentialSoftCapped` method gets called.

File: test/jdk/jdk/internal/util/random/RandomSupportTest.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
+/*
+ * @test
+ * @bug 8326227
+ * @modules java.base/jdk.internal.util.random
+ * @summary Verify that RandomSupport methods behave as specified
```
Comment (jaikiran) [inline test/jdk/jdk/internal/util/random/RandomSupportTest.java:5]: Nit - can you move this `@summary` line after the `@bug` line? It would then match the recommended tag order https://openjdk.org/jtreg/tag-spec.html#ORDER

File: test/jdk/jdk/internal/util/random/RandomSupportTest.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+                havePreviousOutput = true;
+                return Long.MIN_VALUE | 255; // need high value in last byte in order to skip the fast path
+            }
+        }
+    }
+    @Test
+    public void testNextExponentialSoftCapped() {
+        for (double max = 1.0; max < 10.0; max++) {
+            WorstCaseRandomGenerator rng = new WorstCaseRandomGenerator();
+            assertTrue(RandomSupport.computeNextExponentialSoftCapped(rng, max) >= max);
```
Comment (jaikiran) [inline test/jdk/jdk/internal/util/random/RandomSupportTest.java:32]: Would you mind changing this to something like:
```java
double val = RandomSupport.computeNextExponentialSoftCapped(rng, max);
assertTrue(val >= max, val + " isn't >= " + max);

```
That way if this test fails for any reason, then we get the necessary details on what the computed value is.

**Comment by jaikiran** [comment]
> @jddarcy Added a regression test, and currently working on adjusting it (see https://github.com/Pr0methean/jdk/actions/runs/7984444127) to ensure we have a case that fails without the fix, passes with the fix, and is practical to run within the usual unit-test timeouts.

I gave this a try locally. It doesn't fail for me without the source code changes proposed in this PR. I see the following output from the test without the source code changes:

```
got 1.0 for max 1.0
got 2.0 for max 2.0
got 3.0 for max 3.0
got 4.0 for max 4.0
got 5.0 for max 5.0
got 6.0 for max 6.0
got 7.0 for max 7.0
got 11.353912041222094 for max 8.0
got 11.353912041222094 for max 9.0
```

With the proposed changes in this PR, the test continues to pass and I see this output:

```
got 7.569274694148063 for max 1.0
got 7.569274694148063 for max 2.0
got 7.569274694148063 for max 3.0
got 7.569274694148063 for max 4.0
got 7.569274694148063 for max 5.0
got 7.569274694148063 for max 6.0
got 7.569274694148063 for max 7.0
got 11.353912041222094 for max 8.0
got 11.353912041222094 for max 9.0
```

**Comment by Pr0methean** [comment]
@jaikiran Thanks for these results; I'll take another look at the test over the weekend.

File: test/jdk/jdk/internal/util/random/RandomSupportTest.java (hunk 3 above)
Comment (Pr0methean) [reply test/jdk/jdk/internal/util/random/RandomSupportTest.java:4]: Done.

File: src/java.base/share/classes/jdk/internal/util/random/RandomSupport.java (hunk 2 above)
Comment (Pr0methean) [reply src/java.base/share/classes/jdk/internal/util/random/RandomSupport.java:1168]: Updated to return `maxValue` instead.

File: src/java.base/share/classes/jdk/internal/util/random/RandomSupport.java (hunk 1 above)
Comment (Pr0methean) [reply src/java.base/share/classes/jdk/internal/util/random/RandomSupport.java:1176]: Is Guy saying it's impossible for `maxExtraMinus1` to be too small even without this change?

File: test/jdk/jdk/internal/util/random/RandomSupportTest.java (hunk 4 above)
Comment (Pr0methean) [reply test/jdk/jdk/internal/util/random/RandomSupportTest.java:33]: Done.

**Comment by Pr0methean** [comment]
@GuySteele Thanks for your earlier feedback; I've now verified the problem and quantified the rate at which `computeNextExponential` is affected.

**Comment by Pr0methean** [comment]
Update: having reviewed the paper, I'm no longer confident this bug actually exists.
